### PR TITLE
feat: [DHIS2-12337] indicate waiting on saving new tei

### PR DESCRIPTION
**Pull request by jasminenguyennn**

### CODE REVIEW

**Review by superskip — COMMENTED** [review]
Tested the solution with throttled network (1kbit/s) - spinner on save button shows up and cancel button is disabled :+1:
There is no modal showing up when attempting to deselect program while saving :-1:
Save & cancel button gets re-enabled again a few seconds before navigating to enrollment page (not ideal in my opinion).

**Comment by jasminenguyennn** [comment]
@superskip thanks for the review. 
After discussion with @JoakimSM we decided to do it differently: 
When a user clicks the “Save” button and successfully triggers a registration of a new TEI, this button and the “Cancel” should be disabled, and a spinner on the “Save” button should indicate that the request has been sent.
If the user switches context while saving is in progress, cancel the redirect to the to the enrollment overview

So you won't see the modal but the navigation should not be made to enrollment overview either.

**Review by simonadomnisoru — COMMENTED** [review]
Hey @jasminenguyennn , 
Great job on this PR! 🥳 I only have a few small suggestions, hope you can have a look at them.

Firstly to avoid confusion, can you please update both the Jira ticket and the [design docs](https://docs.google.com/document/d/1E5dBP5_BZZ0pLt573Xd0GQB6kve2WUpNdoew3kCFnUo/edit#heading=h.okpfp91vzmyo) with the new feature specs you discussed with Joakim?

> If the user switches context while saving is in progress, cancel the redirect to the to the enrollment overview

Secondly, the possibleDuplicate checks are also performed when a TEI is created without enrollment in a program. Does it make sense to add the same waiting logic for this scenario (in `TeiRegistrationEntry.container`, `TeiRegistrationEntry.component`, `completeSavingNewTrackedEntityInstanceEpic` etc.)?

Screenshot 2023-02-08 at 12 02 49

Thank you!

File: src/core_modules/capture-core/components/ScopeSelector/ScopeSelector.container.js [hunk 1]
```diff
@@ -29,6 +29,7 @@ export const ScopeSelector: ComponentType<OwnProps> = ({
     onResetCategoryOption,
     onResetAllCategoryOptions,
     children,
+    ...restProps
```
[the remark this answers is not in the document]
Comment (jasminenguyennn) [reply src/core_modules/capture-core/components/ScopeSelector/ScopeSelector.container.js:32]: thanks @simonadomnisoru, `isSavingInProgress` and `onContextSwitch` were not used in this container so I think it's okay to use restProps

**Review by simonadomnisoru — APPROVED** [review]

**Review by geethaalwan — APPROVED** [review]
Tested successfully on 2,40,2.39.2,2.38.4 versions

**Review by JoakimSM — COMMENTED** [review]
@jasminenguyennn This got a bit complicated and will be hard to maintain. We are adding a lot of new logic in the generic scope selector that we will have to think about when we are changing it (and reason about when we are looking into the code). We should rather:

- When we start the saving, we should generate a random id (uuid) that will be passed on to the action and also saved in the Redux store (my suggestion would be in the new page, at least for now). When the saving is completed by the offline framework, in the epic that handles the save completion and eventually redirects to the enrollment dashboard, we should verify that the action uuid is still the same as the uuid that is stored in the Redux store before we redirect.
- When the page that contains the form is unmounted we should clear the uuid from the Redux store (meaning that the action uuid and the Redux store uuid won't match after this point and the redirect should not happen)

This way we handle all this logic in the new page and won't add new logic to the scope selector.

**Review by JoakimSM — APPROVED** [review]
Looks good @jasminenguyennn!

**Pull request merged**